pallets/handles/src/lib.rs

```
@@ -0,0 +1,634 @@
+			// Validation: The MSA must not already have a handle associated with it
+			ensure!(
+				MSAIdToDisplayName::<T>::try_get(delegator_msa_id).is_err(),
+				Error::<T>::MSAHandleAlreadyExists
+			);
+
+			// Convert base handle to UTF-8 string slice while validating.
+			let base_handle_str = core::str::from_utf8(&payload.base_handle)
+				.map_err(|_| Error::<T>::InvalidHandleEncoding)?;
+
+			// Validation: The handle length must be valid.
+			// Note: the count() can panic but won't because the base_handle byte length is already checked
+			let len = base_handle_str.chars().count() as u32;
+
+			// Validation: Handle character length must be within range
+			ensure!(
+				len >= HANDLE_BASE_CHARS_MIN && len <= HANDLE_BASE_CHARS_MAX,
+				Error::<T>::InvalidHandleCharacterLength
+			);
+
+			// Validation: The handle must consist of characters not containing reserved words or blocked characters
+			let handle_validator = HandleValidator::new();
+			ensure!(
+				handle_validator.consists_of_supported_unicode_character_sets(base_handle_str),
+				Error::<T>::HandleDoesNotConsistOfSupportedCharacterSets
+			);
+			ensure!(
+				!handle_validator.is_reserved_handle(base_handle_str),
+				Error::<T>::HandleIsNotAllowed
+			);
+			ensure!(
+				!handle_validator.contains_blocked_characters(base_handle_str),
+				Error::<T>::HandleContainsBlockedCharacters
+			);
+
+			// Convert base display handle into a canonical display handle
+			let handle_converter = HandleConverter::new();
+			let canonical_handle_str = handle_converter.convert_to_canonical(&base_handle_str);
+			let canonical_handle_vec = canonical_handle_str.as_bytes().to_vec();
+			let canonical_handle: Handle = canonical_handle_vec.try_into().unwrap();
+
+			// Generate suffix from the next available index into the suffix sequence
+			let suffix_sequence_index =
+				Self::get_next_suffix_index_for_canonical_handle(canonical_handle.clone())
+					.unwrap_or_default();
```

`unwrap_or_default` looks dangerous here.

---

It should ideally do a 0 if unwrap fails @rlaferla  maybe we do a "?" instead of "unwrap_or_default and throw an error instead